compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs

```diff
@@ -1955,6 +1964,410 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                        .into_iter()
+                        .filter_map(|index| struct_def_generics.params.get(index))
+                        .collect();
+
+                    // The struct being constructed is the same as the struct in the impl.
+
+                    let blameable_field_defs: Vec<&ty::FieldDef> = struct_variant_def
+                        .fields
+                        .iter()
+                        .filter(|field_def| {
+                            let field_type: Ty<'tcx> = self.tcx.type_of(field_def.did);
+
+                            // Only retain fields which mention the blameable generics.
+                            struct_def_generics
+                                .iter()
+                                .any(|param| find_param_def_in_ty_walker(field_type.walk(), param))
+                        })
+                        .collect();
+
+                    if blameable_field_defs.len() != 1 {
+                        // We must have a single blameable field, in order to be able to blame it.
+                        return Err(expr);
+                    }
+
+                    let blameable_field_def: &ty::FieldDef = blameable_field_defs[0];
+                    let blameable_field_index = struct_variant_def
+                        .fields
+                        .iter()
+                        .position(|field_def| field_def == blameable_field_def)
+                        .expect("exists in list");
+
+                    if blameable_field_index < call_ctor_args.len() {
+                        return Ok(&call_ctor_args[blameable_field_index]);
+                    }
+
+                    // We failed to find a matching field in the original struct expression.
+                    Err(expr)
+                }
+                _ => Err(expr),
+            },
+            hir::ExprKind::Tup(elements) => {
+                if let &[single_index] = relevant_ty_args_indices.as_slice() {
+                    if single_index < elements.len() {
+                        return Ok(&elements[single_index]);
+                    }
+                }
+                return Err(expr);
+            }
+            _ => Err(expr),
+        }
+    }
+
+    fn point_at_specific_expr_if_possible_for_predicate_obligation(
+        &self,
+        obligation: &PredicateObligation<'tcx>,
+        expr: &'tcx hir::Expr<'tcx>,
+    ) -> Result<&'tcx hir::Expr<'tcx>, &'tcx hir::Expr<'tcx>> {
+        self.point_at_specific_expr_if_possible_for_obligation_cause_code(
+            obligation.cause.code(),
+            expr,
+        )
+    }
+
+    /**
+     * Recursively searches for the most-specific blamable expression.
+     * Specifically, looks for constructor expressions, e.g. `(false, 5)` constructs a tuple
+     * or `Some(5)` constructs an `Option<i32>` and pairs them with their derived impl predicates.
+     */
+    fn point_at_specific_expr_if_possible(
```

Comment:
This is a very dumb request, but do you mind inverting the order you define these functions? This is the entrypoint function, but it comes last, so it reads a bit awkwardly since you're introducing so much code.

You can do this after all the other review tho.